[WIP] Migrating F1 Score

SSaishruthi opened the pull request:
This PR is for migrating F1 score from addons to the core.

@pavithrasv will be reviewing the PR
@seanpmorgan @karmel  for your reference

Commit:
adding f1 code

SSaishruthi commented:
Did not add test scripts yet. Will do after the review comments

SSaishruthi commented:
Hi @pavithrasv 
Did you get a chance to review the PR?

pavithrasv commented on tensorflow/python/keras/metrics.py:
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
+       num_classes : Number of unique classes in the dataset.
+       average : Type of averaging to be performed on data.
+                   Acceptable values are None, micro, macro and
+                   weighted.
```
Are these strings? Could we use quotes around strings for all the usages to make it clear?

pavithrasv commented on tensorflow/python/keras/metrics.py:
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
+    Args:
+       num_classes : Number of unique classes in the dataset.
+       average : Type of averaging to be performed on data.
+                   Acceptable values are None, micro, macro and
```
Could we indent these 2 spaces from where average starts? Also for other usages below.

pavithrasv commented on tensorflow/python/keras/metrics.py:
````diff
@@ -2856,3 +2856,325 @@ def get(identifier):
+    ```python
+    actuals = tf.constant([[1, 1, 0],[1, 0, 0]],
+              dtype=tf.int32)
+    predis = tf.constant([[1, 0, 0],[1, 0, 1]],
````
In general there are a lot of formatting issues, please wrap to next line if > 80 chars and indent continuation line using 4 spaces.

pavithrasv commented on tensorflow/python/keras/metrics.py:
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
+             dtype=tf.int32)
+    # F-Beta Micro
+    fb_score = tfa.metrics.FBetaScore(num_classes=3,
+                beta=0.4, average='micro')
```
tf.metrics

pavithrasv commented on tensorflow/python/keras/metrics.py:
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
+        if beta <= 0.0:
+            raise ValueError("beta value should be greater than zero")
+        else:
+            self.beta = beta
```
Please use ' for quotation to be consistent with the rest of the file.

pavithrasv commented on tensorflow/python/keras/metrics.py:
```diff
@@ -2856,3 +2856,325 @@ def get(identifier):
+
+    # TO DO SSaishruthi: Add sample weight option
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        y_true = tf.cast(y_true, tf.int32)
```
Please raise an exception for unsupported use cases.

pavithrasv reviewed (commented):
Thank you for the PR. Did a quick review, in general i am seeing a lot of formatting issues, please refer to the code in the rest of the file for formatting details and also please add comprehensive unit tests for the metric.

SSaishruthi commented:
Thanks @pavithrasv 
Do you have any sample test scripts that I can refer to? I have written one for addons and will combine both.

Commit:
addressing comments

Commit:
addressing comments

SSaishruthi commented:
Hi @pavithrasv 

I have updated the following:

1.  Corrected lines > 80 characters.
2. Changed all `tf.math` to `math_ops` and `tf.reduce_sum` to `math_ops.reduce_sum`
3. Added usage for Keras API
4. Changed `tfa.metrics` to `tf.keras.metrics`.
5. Changed indentation for docstrings

Clarification:

1. We are not exposing `sample_weight` parameter now. Do we need to add exception for the same?
2. I have test cases for F1 in addons. Can we modify the same or I need to refer some existing sample scripts?

https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/f1_test.py


Please let me know if the current changes look good

SSaishruthi commented:
I will be adding a `threshold` parameter today. Will ping again once I update

PhilipMay commented:
There are still issues in the implementation. See here: https://github.com/tensorflow/addons/issues/490

SSaishruthi commented:
@pavithrasv The issue has been taken care. we are good for now

SSaishruthi commented:
Updating scripts

SSaishruthi commented:
Waiting for the f1 score update PR to get merged. Will update the scripts right after that

SSaishruthi commented:
We have an updated version of the code. It is almost ready. I will update the code here by tomorrow and we can take it forward from here.

Commit:
new scripts

SSaishruthi commented:
Refractor PR got merged. Now, we are good to go with migration.
Updated new scripts. Need to do some minor formatting work.

pavithrasv commented:
You can add tests in the file metrics_test.py

SSaishruthi commented:
Thanks, @pavithrasv will update.

Commit:
adding test

SSaishruthi commented:
I have added a test. Will do a final cleanup in a day and we can have a final review on this.
Thanks

tensorflowbutler commented:
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

SSaishruthi commented:
This is WIP. Will be completed this week

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

PhilipMay commented:
F1 metric in addons still seems to have a very ugly bug: https://github.com/tensorflow/addons/issues/825

gbaned commented:
@SSaishruthi Any update on this PR, please. Thanks!

SSaishruthi commented:
@gbaned Hi

Sorry I was on vacation. We have some clarification to make. Will keep till we clear that

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
@SSaishruthi Any update on this PR and can you please resolve conflicts? Thanks!

PhilipMay commented:
@gbaned No offence but I think @SSaishruthi does not have the time at the moment. Someone else should pick up the migration work. The TF Addons implementation seems to be ok. Want me to create a new PR for this migration?

gbaned commented:
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
